### ISSUE TITLE
user paths will always be used even if they don't exist

### ISSUE THREAD

**Issue by dmiller423** [reicast/reicast-emulator#1398]
@luserx0 : How about you handle this?   Simple C++ path checking

</ Everyone else please ignore horribly formatted issue >

!LoadRomFiles(get_readonly_data_path(DATA_PATH)))

> 

string get_writable_data_path(const string& filename)
{
	/* Only stuff in the user_data_dir is supposed to be writable,
	 * so we always return that.
	 */
	return (user_data_dir + filename);
}
string get_readonly_data_path(const string& filename)
{
	string user_filepath = get_writable_data_path(filename);
	if(file_exists(user_filepath))
	{
		return user_filepath;
	}

	string filepath;
	for (unsigned int i = 0; i < system_data_dirs.size(); i++) {
		filepath = system_data_dirs[i] + filename;
		if (file_exists(filepath))
		{
			return filepath;
		}
	}

	// Not found, so we return the user variant
	return user_filepath;
}

calls file_exists on directories instead of the files,  ok to check if directory exists but the files are prob what this is meant for... 

the bigger issue is :

	SaveRomFiles(get_writable_data_path("/data/"));

Which will always return (user_data_dir + "/data/");

even if user_data_dir is empty ...  

void SaveRomFiles(const string& root)
{
	sys_nvmem.Save(root, ROM_PREFIX, "nvmem.bin", "nvmem");
}

^ can never be written to system_* paths,  they are useless at this point only the user_data_dir has any meaning... 

get_readonly_data_path() can test user_data_path.empty() then check sys, then likely might want to return ./data instead of  FSROOT/data

get_writable_data_path()  could do similar,  some default system path needs to be added prob.

Most importantly maybe these should be called inside of Load/Save RomFiles per file ( esp for reads ! )

**Comment by luserx0** [comment]
C++17 Introduced this, https://en.cppreference.com/w/cpp/filesystem filesystem lib. Can I use it somehow?

Doing this for Lunix at first (Unix and BSD included) would make sense. I'm not sure we can have a POSIX way of setting defaults, but that's expected.

I'm gonna use the set_user_config_dir, set_user_data_dir, add_system_config_dir, add_system_data_dir functions located on top of stdclass.cpp
As I see on my end, these are implemented but not called at any point, could you confirm this?

I'll have to ask you some questions as well concerning our structure.

### WIP!!

**Comment by baka0815** [comment]
We can't use C++17 currently as AppVeyor still uses VS 2013.

**Comment by dmiller423** [comment]
AppVeyor  is not fixed to VS2013, you can specify ... i think we switched to 2015 as of a few days ago officially since it allows multi-targeting in project file to work in vs2017 as well.   Still, the problem here isn't related really @luserx0 , the problem is that things just aren't done in the right place.

  The code for checking is fine, but check for the files not for the directory,   It checking for dirs with passed DATA_PATH ("/data/") is bad it'll always be a workable path esp for write 👎 

More importantly is the fact the get_writable_data_path() always uses user_dir ...  if user_dir is empty it should check system paths before defaulting to a relative ./data/.

**Comment by baka0815** [comment]
Nope, #1400 was still build with v120 which doesn't know inline and therefore failed at first.

**Comment by AbandonedCart** [comment]
You can change "120" to "140" in appveyor and it builds 2015.

**Comment by baka0815** [comment]
Sorry for OT then: how?
